airbyte-integrations/bases/base-normalization/README.md

```diff
@@ -1,11 +1,147 @@
+* [architecture / Basic Normalization](../../../docs/understanding-airbyte/basic-normalization.md)
 * [tutorials / Custom dbt normalization](../../../docs/operator-guides/transformation-and-normalization/transformations-with-dbt.md)
 
-# Testing normalization
+## Under the hood
+
+Normalization has two Python modules:
+* `transform_config` parses the destination connector config and generates a profile.yml file,
+  which configures how dbt will connect to the destination database.
+* `transform_catalog` parses the connection's catalog and generates a dbt_project.yml file,
+  which configures the models that dbt will run and how they should be materialized.
+
+`entrypoint.sh` (the entrypoint to normalization's Docker image) invokes these two modules, then calls `dbt run` on their output.
+
+### Incremental updates
+
+When generating the final table, we need to pull data from the SCD model.
+A naive implementation would require reading the entire SCD table and completely regenerating the final table on each run.
+This is obviously inefficient, so we instead use dbt's [incremental materialization mode](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models).
+At each stage of the dbt pipeline, normalization will query the target table for the newest `_airbyte_emitted_at` value.
+Then we only need to find records from the source table with `_airbyte_emitted_at` greater than or equal to that value
+(equal to is necessary in case a previous normalization run was interrupted).
+
+This handles the two error scenarios quite cleanly:
+* If a sync fails but succeeds after a retry, such that the first attempt commits some records and the retry commits a superset
+  of those records, then normalization will see that the SCD table has none of those records. The SCD model has a deduping stage,
+  which removes the records which were synced multiple times.
+* If normalization fails partway through, such that (for example) the SCD model is updated but the final table is not, and then the sync
+  is retried, then the source will not re-emit any old records (because the destination will have emitted a state message ack-ing
+  all of the records). If the retry emits some new records, then normalization will append them to the SCD table as usual
+  (because, from the SCD's point of view, this is just a normal sync). Then the final table's latest `__airbyte_emitted_at`
+  will be older than the original attempt, so it will pull both the new records _and_ the first attempt's records from the SCD table.
+
+## Developer workflow
+
+At a high level, this is the recommended workflow for updating base-normalization:
+1. Manually edit the models in `integration_tests/normalization_test_output/postgres/test_simple_streams/models/generated`.
+   Run `dbt compile` and manually execute the SQL queries. This requires manual setup and validation, but allows you to quickly experiment
```

Comment:
I actually never really use `dbt compile`, but yes this step about 
> manually execute the SQL queries
means that you do that in another SQL editor/ui? like your datagrip or bigquery web-ui interactively?

a little slower than doing it interactively / manually in a specialized tool mentioned above, you can also run `dbt run --models <name of the table>` to experiment with the SQL from dbt.

Reply:
interesting - I actually preferred using compile. Maybe just because I kept making syntax errors, so I needed to edit the compiled sql a few times to get it right. Will add the `dbt run` version to the doc though.

Reply:
but yeah, I was manually messing around in the CLI/UI (psql, bigquery web-ui) a lot. Was also useful for running `explain` queries or just visually inspecting the results.